tests: Consolidate stream access tests into table-driven structure.

Git-HimanshuRathi opened the pull request:
This PR refactors the stream access tests in `stream_data.test.cjs` by replacing the separate
`has_metadata_access`, `has_content_access`, and `can_preview` test blocks with a single,
cleaner table-driven test.

### Why this change?

As described in issue #33600:

- The previous tests repeated large amounts of setup code.
- Behavior differences between the access checks were not easy to compare.
- Zulip prefers a table-driven pattern for test clarity (e.g., `filter.test.cjs`).

This update applies that approach to the stream access tests.

### What this PR does

- Introduces a unified table-driven test (`"stream access table-driven test"`).
- Tests combinations of:
  - public / private / web-public streams
  - subscribed vs. unsubscribed user states
  - admin / moderator / guest / normal user roles
  - spectator behavior
  - history_public_to_subscribers conditions
- Asserts all three access functions together:
  - `has_metadata_access(stream)`
  - `has_content_access(stream)`
  - `can_preview(stream)`

This removes duplication and significantly improves readability and maintainability.

### Test results

All relevant tests passed locally:
✔ ./tools/test-js-with-node web/tests/stream_data.test.cjs — passed
✔ No regressions in other JS tests
✔ Backend + webhook tests unaffected (no backend code touched)
### Issue

Fixes: #33600

<details>
<summary><b>Self-review checklist</b></summary>

- [x] Self-reviewed the changes for clarity and maintainability.
- [x] Communicate decisions, questions, and potential concerns.
- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.
- [x] Individual commits are ready for review (see commit discipline).
- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.
- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

</details>

alya commented:
Thanks! Please fill out the self-review checklist, which is for some reason missing from the PR description.

Your PR will also need to pass tests.

Git-HimanshuRathi commented:
@alya 

> Thanks! Please fill out the self-review checklist, which is for some reason missing from the PR description.
> 
> Your PR will also need to pass tests.

I’ve now added the full self-review checklist to the PR description, as requested.
I also pushed a new commit fixing the previously failing CI test case.

Git-HimanshuRathi commented on web/tests/stream_data.test.cjs:
```diff
@@ -2012,295 +1959,127 @@ test("options for dropdown widget", () => {
-    assert.equal(stream_data.has_metadata_access(social), true);
+        user: {is_admin: false, is_guest: false},
+        spectator: false,
+        expect: {meta: true, content: true, preview: false},
```
done

Git-HimanshuRathi commented:
@shubham-padia  

Could you please take a look and let me know if there’s anything that needs to be changed ?

shubham-padia commented on web/tests/stream_data.test.cjs:
```diff
@@ -2303,58 +2413,79 @@ run_test("can_toggle_subscription", ({override}) => {
+    assert.equal(stream_data.can_access_stream_email(social), false);
+
+    override(current_user, "is_guest", true);
+    assert.equal(stream_data.can_access_stream_email(social), false);
```
How are you deciding which tests are part of the accest_test_cases  and which are not

shubham-padia reviewed (commented):
I've left a few comments, but the diff is hard to parse. Please do not change locations of tests you haven't changed.  From what I can see `can_access_stream_email` is unchanged but it's location has been changed, so it looks like you changed that code. 

Please make sure any refactors you are doing are their own commit. If possible break down the changes into multiple coherent commits so that easier to review.

Git-HimanshuRathi commented:
> I've left a few comments, but the diff is hard to parse. Please do not change locations of tests you haven't changed. From what I can see `can_access_stream_email` is unchanged but it's location has been changed, so it looks like you changed that code.
> 
> Please make sure any refactors you are doing are their own commit. If possible break down the changes into multiple coherent commits so that easier to review.

Hi @shubham-padia,

Thank you for the feedback. I've addressed the missing test cases (including the guest + unsubscribed = false case you pointed out).

However, I understand the PR still has reviewability concerns:

Multiple types of changes mixed together
Tests being moved/rewritten
Single large commit instead of separated commits
I'm thinking of restructuring into separate commits:

Commit 1: Table-driven refactor only (replacing has_metadata_access, 
has_content_access
, can_preview with unified access_cases)
Commit 2: New standalone tests for group permissions (stream access with group permissions, can_subscribe_group grants access, etc.)
Commit 3: Any restorations or fixes to existing tests
Would this approach work, or would you prefer a different structure?

Thanks!

Git-HimanshuRathi commented:
Hi @shubham-padia,

I've restructured the PR based on your feedback. Here's what changed:

Structure:

Single commit with table-driven refactor
Consolidated all stream access tests into access_cases (15 test cases)
Group permissions (can_add_subscribers_group, can_subscribe_group, can_administer_channel_group) are now part of the table
Removed redundant standalone tests that duplicated table coverage
Kept only can_preview with content access override as standalone (uses rewire pattern)

What I did NOT change:

can_toggle_subscription — remains in its original location
can_access_stream_email — remains in its original location

Git-HimanshuRathi commented on web/tests/stream_data.test.cjs:
```diff
@@ -2271,19 +2383,17 @@ run_test("can_toggle_subscription", ({override}) => {
 
-    social.subscribed = false;
-    assert.equal(stream_data.can_toggle_subscription(social), false);
-
```
added again . removed it by mistake .

shubham-padia reviewed (commented):
Thanks, good work, the diff is much clearer to parse now.

 What I encourage you next to focus on is the naming of the tests, the name of each case does not tell me what the test does on the first glance. Keep the title short but clear. `public channel unsubscribed normal user` is just 5 words bunched together, but does not tell me what scenario are we specifically testing for. Are we testing specally for metadata access, for content access or for both. The title should be valid plain english. 

In addition to that, I found a missing test case from a quick glance. Now that the diff is more clear, you should be able to go through each line of the old test and see that the corresponding test scenario exists in access_cases.

Git-HimanshuRathi commented:
> Thanks, good work, the diff is much clearer to parse now.
> 
> What I encourage you next to focus on is the naming of the tests, the name of each case does not tell me what the test does on the first glance. Keep the title short but clear. `public channel unsubscribed normal user` is just 5 words bunched together, but does not tell me what scenario are we specifically testing for. Are we testing specally for metadata access, for content access or for both. The title should be valid plain english.
> 
> In addition to that, I found a missing test case from a quick glance. Now that the diff is more clear, you should be able to go through each line of the old test and see that the corresponding test scenario exists in access_cases.

@shubham-padia Thanks for the feedback! Since the diff is now clearer, I've addressed both issues:

1. Test naming — Updated all 19 test case descriptions to use plain English that clearly describes the expected behavior.

2. Missing test case — Added the case for "guest in can_administer_channel_group is denied metadata access to private channel".

I also added matching cases for guests in can_add_subscribers_group and can_subscribe_group for completeness, since guests should be denied access regardless of which group permission they have.

Please let me know if anything else needs adjustment!

Git-HimanshuRathi commented:
@shubham-padia I have made the changes as requested . can you review it please ?

shubham-padia commented on web/tests/stream_data.test.cjs:
```diff
@@ -2057,192 +2057,291 @@ test("can_access_stream_email", ({override}) => {
+        expect: {has_metadata_access: true, has_content_access: true, can_preview: false},
+    },
+    {
+        desc: "normal user has full access to web-public channel",
```
`normal` is a vague term, `unsubscribed` would be better here.

timabbott commented:
Merged, thanks @Git-HimanshuRathi and @shubham-padia! 

This does feel a lot more robust. If you're looking for a next project, you could try doing something similar with some of the other tests with a similar set of assertions in the structure; I would check the `user_group` tests and `filter.test.cjs`  for having a similar set of problems.
